File: typescript/src/agentOverlapAnalyzer.ts
import { TfIdf } from "natural";
import { removeStopwords } from "stopword";
import { Logger } from "./utils/logger";

export interface OverlapResult {
  overlapPercentage: string;
  potentialConflict: "High" | "Medium" | "Low";
}

export interface UniquenessScore {
  agent: string;
  uniquenessScore: string;
}

export interface AnalysisResult {
  pairwiseOverlap: { [key: string]: OverlapResult };
  uniquenessScores: UniquenessScore[];
}

export class AgentOverlapAnalyzer {
  private agents: { [key: string]: { name: string; description: string } };

  constructor(agents: {
    [key: string]: { name: string; description: string };
  }) {
    this.agents = agents;
  }

  analyzeOverlap(): void {
    const agentDescriptions = Object.entries(this.agents).map(
      ([_, agent]) => agent.description
    );
    const agentNames = Object.entries(this.agents).map(([key, _]) => key);

    if (agentNames.length < 2) {
      Logger.logger.info("Agent Overlap Analysis requires at least two agents.");
      Logger.logger.info(`Current number of agents: ${agentNames.length}`);
      if (agentNames.length === 1) {
        Logger.logger.info(`\nSingle Agent Information:`);
        Logger.logger.info(`Agent Name: ${agentNames[0]}`);
        Logger.logger.info(`Description: ${agentDescriptions[0]}`);
      }
      return;
    }

    
    const tfidf = new TfIdf();

    // Preprocess descriptions and add to TF-IDF
    const _preprocessedDescriptions = agentDescriptions.map((description) => {
      const tokens = removeStopwords(description.toLowerCase().split(/\W+/));
      tfidf.addDocument(tokens);
      return tokens;
    });

    const overlapResults: { [key: string]: OverlapResult } = {};
    for (let i = 0; i < agentDescriptions.length; i++) {
      for (let j = i + 1; j < agentDescriptions.length; j++) {
        const agent1 = agentNames[i];
        const agent2 = agentNames[j];
        const similarity = this.calculateCosineSimilarity(
          tfidf.listTerms(i),
          tfidf.listTerms(j)
        );
        const overlapPercentage = (similarity * 100).toFixed(2);
        const key = `${agent1}__${agent2}`;
        overlapResults[key] = {
          overlapPercentage: `${overlapPercentage}%`,
          potentialConflict:
            similarity > 0.3 ? "High" : similarity > 0.1 ? "Medium" : "Low",
        };
      }
    }

    // Calculate uniqueness scores
    const uniquenessScores: UniquenessScore[] = agentDescriptions.map(
      (description, index) => {
        const otherDescriptions = agentDescriptions.filter(
          (_, i) => i !== index
        );
        const similarities = otherDescriptions.map((otherDescription) => {
          const key1 = `${agentNames[index]}__${agentNames[otherDescriptions.indexOf(otherDescription)]}`;
          const key2 = `${agentNames[otherDescriptions.indexOf(otherDescription)]}__${agentNames[index]}`;
          const result = overlapResults[key1] || overlapResults[key2];
          return result ? parseFloat(result.overlapPercentage) / 100 : 0;
        });
        const avgSimilarity =
          similarities.reduce((sum, sim) => sum + sim, 0) / similarities.length;
        return {
          agent: agentNames[index],
          uniquenessScore: ((1 - avgSimilarity) * 100).toFixed(2) + "%",
        };
      }
    );

    // Print pairwise overlap results
    Logger.logger.info("Pairwise Overlap Results:");
    Logger.logger.info("_________________________\n");
    for (const key in overlapResults) {
      const [agent1, agent2] = key.split("__");
      const { overlapPercentage, potentialConflict } = overlapResults[key];
      Logger.logger.info(
        `${agent1} - ${agent2}:\n- Overlap Percentage - ${overlapPercentage}\n- Potential Conflict - ${potentialConflict}\n`
      );
    }
    Logger.logger.info("");

    // Print uniqueness scores
    Logger.logger.info("Uniqueness Scores:");
    Logger.logger.info("_________________\n");
    uniquenessScores.forEach((score) => {
      Logger.logger.info(
        `Agent: ${score.agent}, Uniqueness Score: ${score.uniquenessScore}`
      );
    });
  }

  private calculateCosineSimilarity(
    terms1: { term: string; tfidf: number }[],
    terms2: { term: string; tfidf: number }[]
  ): number {
    const vector1: { [term: string]: number } = {};
    const vector2: { [term: string]: number } = {};
    terms1.forEach((term) => (vector1[term.term] = term.tfidf));
    terms2.forEach((term) => (vector2[term.term] = term.tfidf));

    const terms = new Set([...Object.keys(vector1), ...Object.keys(vector2)]);
    let dotProduct = 0;
    let magnitude1 = 0;
    let magnitude2 = 0;

    for (const term of terms) {
      const v1 = vector1[term] || 0;
      const v2 = vector2[term] || 0;
      dotProduct += v1 * v2;
      magnitude1 += v1 * v1;
      magnitude2 += v2 * v2;
    }

    magnitude1 = Math.sqrt(magnitude1);
    magnitude2 = Math.sqrt(magnitude2);

    if (magnitude1 && magnitude2) {
      return dotProduct / (magnitude1 * magnitude2);
    } else {
      return 0;
    }
  }
}

File: python/src/agent_squad/agent_overlap_analyzer.py
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from typing import Dict, List
from agent_squad.utils.logger import Logger
import numpy as np
import nltk
from nltk.corpus import stopwords
import re

nltk.download('stopwords')
STOPWORDS = set(stopwords.words('english'))


class OverlapResult:
    def __init__(self, overlap_percentage: str, potential_conflict: str):
        self.overlap_percentage = overlap_percentage
        self.potential_conflict = potential_conflict


class UniquenessScore:
    def __init__(self, agent: str, uniqueness_score: str):
        self.agent = agent
        self.uniqueness_score = uniqueness_score


class AgentOverlapAnalyzer:
    def __init__(self, agents: Dict[str, Dict[str, str]]):
        if not agents:
            raise ValueError("Agents dictionary cannot be empty.")
        self.agents = agents

    def analyze_overlap(self):
        agent_descriptions = [self._preprocess(agent["description"]) for agent in self.agents.values()]
        agent_names = list(self.agents.keys())

        if len(agent_names) < 2:
            raise ValueError("Agent Overlap Analysis requires at least two agents.")  # Raise ValueError
        tfidf_matrix = self._build_tfidf_matrix(agent_descriptions)
        overlap_results = self.calculate_pairwise_overlap(agent_names, tfidf_matrix)
        uniqueness_scores = self.calculate_uniqueness_scores(agent_names, tfidf_matrix)
        self.log_results(overlap_results, uniqueness_scores)

    def _preprocess(self, text: str) -> str:
        tokens = re.findall(r'\b\w+\b', text.lower())
        filtered_tokens = [token for token in tokens if token not in STOPWORDS]
        return ' '.join(filtered_tokens)

    def _build_tfidf_matrix(self, agent_descriptions: List[str]):
        vectorizer = TfidfVectorizer()
        return vectorizer.fit_transform(agent_descriptions)

    def calculate_pairwise_overlap(self, agent_names: List[str], tfidf_matrix) -> Dict[str, OverlapResult]:
        overlap_results = {}
        for i in range(len(agent_names)):
            for j in range(i + 1, len(agent_names)):
                agent1 = agent_names[i]
                agent2 = agent_names[j]
                similarity = cosine_similarity(tfidf_matrix[i], tfidf_matrix[j])[0][0]
                overlap_percentage = f"{similarity * 100:.2f}%"
                key = f"{agent1}__{agent2}"
                overlap_results[key] = OverlapResult(
                    overlap_percentage=overlap_percentage,
                    potential_conflict="High" if similarity > 0.3 else "Medium" if similarity > 0.1 else "Low"
                )
        return overlap_results

    def calculate_uniqueness_scores(self, agent_names: List[str], tfidf_matrix) -> List[UniquenessScore]:
        uniqueness_scores = []
        for index in range(len(agent_names)):
            similarities = [
                cosine_similarity(tfidf_matrix[index], tfidf_matrix[i])[0][0]
                for i in range(len(agent_names)) if i != index
            ]
            avg_similarity = np.mean(similarities) if similarities else 0
            uniqueness_score = f"{max(0, (1 - avg_similarity)) * 100:.2f}%"  # Ensure non-negative score
            uniqueness_scores.append(UniquenessScore(
                agent=agent_names[index],
                uniqueness_score=uniqueness_score
            ))
        return uniqueness_scores

    def log_results(self, overlap_results: Dict[str, OverlapResult], uniqueness_scores: List[UniquenessScore]):
        Logger.info("Pairwise Overlap Results:")
        Logger.info("_________________________\n")
        for key, value in overlap_results.items():
            agent1, agent2 = key.split("__")
            Logger.info(f"{agent1} - {agent2}:\n- Overlap Percentage - {value.overlap_percentage}\n- Potential Conflict - {value.potential_conflict}\n")

        Logger.info("Uniqueness Scores:")
        Logger.info("_________________\n")
        for score in uniqueness_scores:
            Logger.info(f"Agent: {score.agent}, Uniqueness Score: {score.uniqueness_score}")
